### PR TITLE
Correct unit tests for Environment.SetEnvironmentVariable

### LINKED ISSUE #16766: System.Environment rejects env variable values acceptable to the OS
In https://github.com/dotnet/coreclr/issues/9793 an environment variable already set in the OS caused an ArgumentException when MSBuild tried to set it because the length of the value was over 32KB. (CI apparently passes the change description in an environment variable)

We should accept any environment variable that's acceptable to the OS. Apparently on Windows Server 2012 (this run) over 32KB is acceptable. Rather than check the length we should perhaps let Windows return an error.

https://github.com/dotnet/coreclr/blob/master/src/mscorlib/src/System/Environment.cs#L45
https://github.com/dotnet/coreclr/blob/master/src/mscorlib/src/System/Environment.cs#L843

### PR DESCRIPTION
**Pull request by yvanin**
Fix #16766

### CODE REVIEW

File: src/System.Runtime.Extensions/tests/System/Environment.SetEnvironmentVariable.cs [hunk 1]
```diff
@@ -31,9 +30,53 @@ public void IncorrectVariableThrowsArgument()
...
+            finally
+            {
+                Environment.SetEnvironmentVariable(var, null);
+            }
+        }
+
+        [Fact]
+        public void EnvironmentVariableTooLarge_Throws()
+        {
+            string var = "Test_SetEnvironmentVariable_EnvironmentVariableTooLarge_Throws";
```
Comment (danmosemsft) [inline src/System.Runtime.Extensions/tests/System/Environment.SetEnvironmentVariable.cs]: What about an excessively long environment variable name? We should have a test for it, whatever it does.

**Comment by yvanin** [comment]
I made the tests that catch OutOfMemoryException to run on Windows only.
I cannot reproduce the same behavior on other OSes because the string constructor fails first.

**Comment by danmosemsft** [comment]
@weshaggard this is for the change you had to fix tests for

**Comment by stephentoub** [comment]
@dotnet-bot test this please

**Comment by yvanin** [comment]
There are some failed unit tests unrelated to these changes but also my tests are failing in some environments but not the others.

For instance, no exception is thrown when OutOfMemoryException is expected on _Windows.7.Amd64.Open:Debug-x64_, but it is thrown on Windows 8.1 and 10.

It is problematic since the string long enough to throw OutOfMemoryException in Environment.SetEnvironmentVariable cannot be constructed.

Should I convert the tests to either accept OutOfMemoryException or no exception at all? It doesn't exactly test the required behavior but better than nothing.

**Comment by yvanin** [comment]
The CoreCLR changes haven't propagated to NETFX

**Comment by danmosemsft** [comment]
@yvanin changes do not automatically propagate to NETFX. The code is often diverged, and there is a higher compatibility bar (because it does not ship SXS, and because it has very many established customers). We periodically port selected fixes.

To fix the NETFX tests, please attribute the tests like
`[SkipOnTargetFramework(TargetFrameworkMonikers.NetFramework, ".NET Framework does not have the fix to allow arbitrary length environment variables.")]`

**Comment by danmosemsft** [comment]
@dotnet-bot test Linux arm Release Build please (??)

**Comment by danmosemsft** [comment]
Looks like 32bit can't create the string...

https://mc.dot.net/#/user/yvanin/pr~2Fjenkins~2Fdotnet~2Fcorefx~2Fmaster~2F/test~2Ffunctional~2Fcli~2F/dc1eec6a3d0cc0402da779fcce9ef209db7e0c7f/workItem/System.Runtime.Extensions.Tests/analysis/xunit/System.Tests.SetEnvironmentVariable~2FEnvironmentVariableTooLarge_Throws

**Comment by danmosemsft** [comment]
@yvanin two of the failures, both on Windows 7, seem to be "insufficient reosurces" when an unrelated test is starting a process. My guess is this process is starting during the interval when your concurrently running test has set a ginormous environment variable in the parent test process. To fix this please change those tests that set ginormous environment variables to run in their own process. this is as simple as
- make test class derive from `RemoteExecutorTestBase`
- put test method body inside `RemoteInvoke(() => { ... }`
there are lots of existing examples.

**Comment by yvanin** [comment]
That fixed it

**Comment by danmosemsft** [comment]
Thanks @yvanin. Perhaps there's another issue you'd like to look at? Usually PRs are much smoother after the first couple.

**Pull request merged**